wave_containers/include/wave/containers/measurement_container.hpp

```diff
@@ -0,0 +1,313 @@
+     * not already exist.
+     *
+     * @return a pair p. If and only if insertion occurred, p.second is true and
+     * p.first points to the element inserted.
+     */
+    std::pair<iterator, bool> insert(const MeasurementType &);
+
+    /** Insert a Measurement constructed from the arguments if a measurement for
+     * the same time and sensor does not already exist.
+     *
+     * @return a pair p. If and only if insertion occurred, p.second is true and
+     * p.first points to the element inserted.
+     */
+    template<typename... Args>
+    std::pair<iterator, bool> emplace(Args &&... args);
+
+    /** Delete the element with the matching time and sensor id, if one exists.
+     *
+     * @return the number of elements deleted.
+     */
+    size_type erase(const TimeType &, const SensorIdType &);
+
+    /** Delete all elements */
+    void clear() noexcept;
+
+    // Retrieval
+
+    /** Get the value of a measurement with corresponding time and sensor id */
+    const ValueType get(const TimeType &, const SensorIdType &) const;
+
+    // Iterators
+
+    iterator begin() noexcept;
+    iterator end() noexcept;
+    const_iterator begin() const noexcept;
+    const_iterator end() const noexcept;
+    const_iterator cbegin() const noexcept;
+    const_iterator cend() const noexcept;
+
+ private:
+    using composite_type = typename
+    internal::measurement_container<T>::composite_type;
+
+    // Helper to get the composite index
+    composite_type &composite() noexcept;
+    const composite_type &composite() const noexcept;
+
+    // Internal multi_index_container
+    typename internal::measurement_container<T>::type storage;
+};
+
+template<typename T>
+MeasurementContainer<T>::MeasurementContainer() {
+
+}
+
+template<typename T>
+std::pair<typename MeasurementContainer<T>::iterator, bool>
+MeasurementContainer<T>::insert(const MeasurementType &m) {
+    return this->composite().insert(m);
+}
+
+template<typename T>
+template<typename... Args>
+std::pair<typename MeasurementContainer<T>::iterator, bool>
+MeasurementContainer<T>::emplace(Args &&... args) {
+    // Support Boost.MultiIndex <= 1.54, which does not have emplace()
+#if BOOST_VERSION < 105500
+    return this->composite().insert(
+            MeasurementType{std::forward<Args>(args)...});
+#else
+    return this->composite().emplace(std::forward<Args>(args)...);
+#endif
+}
+
+template<typename T>
+typename MeasurementContainer<T>::size_type
+MeasurementContainer<T>::erase(const TimeType &t, const SensorIdType &s) {
+    auto &composite = this->composite();
+    auto it = composite.find(boost::make_tuple(s, t));
+    if (it == composite.end()) {
+        return 0;
+    }
+    composite.erase(it);
+    return 1;
+}
+
+template<typename T>
+const typename MeasurementContainer<T>::ValueType
+MeasurementContainer<T>::get(const TimeType &t, const SensorIdType &s) const {
+    const auto &composite = this->composite();
```

Comment:
Yes. As I suggested in your PR, `const auto &` makes it clear that's what I'm doing.